File: sdk/typescript/src/resources/audit.ts
import { HttpClient } from "../http.js";
import type {
  AuditChainVerification,
  AuditList,
  AuditListParams,
  AuditStats,
} from "../types.js";

const BASE = "/api/v1/audit";

export class AuditResource {
  constructor(private readonly http: HttpClient) {}

  /** List audit log entries with optional agent filter. */
  async list(params?: AuditListParams): Promise<AuditList> {
    return this.http.get<AuditList>(BASE, params as Record<string, unknown>);
  }

  /** Get aggregated audit statistics. */
  async stats(agentId?: string): Promise<AuditStats> {
    return this.http.get<AuditStats>(`${BASE}/stats`, { agent_id: agentId });
  }

  /** Verify the HMAC chain integrity for an agent's audit log. */
  async verifyChain(agentId: string): Promise<AuditChainVerification> {
    return this.http.get<AuditChainVerification>(`${BASE}/verify/${agentId}`);
  }
}

File: sdk/python/src/ai_identity/resources/audit.py
"""Audit resource — query logs, stats, and verify chain integrity."""

from ai_identity._http import HTTPClient
from ai_identity.models.audit import AuditChainVerification, AuditList, AuditStats

BASE = "/api/v1/audit"


class AuditResource:
    """Query tamper-proof audit logs and forensic data."""

    def __init__(self, http: HTTPClient) -> None:
        self._http = http

    async def list(
        self,
        *,
        agent_id: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> AuditList:
        """List audit log entries with optional agent filter."""
        params = {"agent_id": agent_id, "limit": limit, "offset": offset}
        resp = await self._http.get(BASE, params=params)
        return AuditList.model_validate(resp)

    async def stats(self, *, agent_id: str | None = None) -> AuditStats:
        """Get aggregated audit statistics."""
        params = {"agent_id": agent_id}
        resp = await self._http.get(f"{BASE}/stats", params=params)
        return AuditStats.model_validate(resp)

    async def verify_chain(self, agent_id: str) -> AuditChainVerification:
        """Verify the HMAC chain integrity for an agent's audit log.

        Returns whether the chain is intact and, if not, where it broke.
        """
        resp = await self._http.get(f"{BASE}/verify/{agent_id}")
        return AuditChainVerification.model_validate(resp)
